iot_verifier/src/tx_scaler.rs

```diff
@@ -89,10 +74,10 @@ impl Server {
             .gateways_recent_activity(refresh_start)
             .await
             .map_err(sqlx::Error::from)?;
-        let mut gw_stream = self.iot_config_client.stream_gateways_info().await?;
-        while let Some(gateway_info) = gw_stream.next().await {
-            if let Some(metadata) = gateway_info.metadata {
-                if active_gateways.contains_key(&gateway_info.address.as_ref().to_vec()) {
+        for k in active_gateways.keys() {
+            let pubkey = PublicKeyBinary::from(k.clone());
+            if let Some(gateway_info) = self.gateway_cache_receiver.borrow().get(&pubkey) {
+                if let Some(metadata) = &gateway_info.metadata {
```

Comment:
the `GatewayMap` in the receiver watch channel is just a hashmap so you can save all of the pubkey cloning and serialization here by just calling `self.gateway_cache_receiver.borrow().values()` and you get an iterator over all of the gateway info entries in the map that you can then match on with `let Some(metadata) = gateway.metadata {...etc...}`. i would imagine with over 400K gateways on average for each run, avoiding all that pubkey deserialization that will significantly improve the performance of this function

Reply:
I agree the vec -> pubkey binary of the pubkey here  is sub optimal, but not sure what you suggest is gonna be much better,  the gateway_cache map contains  *all* gateways ( asserted, unasserted etc and a count of ~980k ) whilst the active_gateways maps is obviously only those recently active ( <400k ).

I would need to iterative over the 980k cached gateways in the approach you suggest rather than the smaller active list.

Reply:
It’d be interesting to benchmark both solutions but since we’re not talking about that much data per gateway (these are pretty tiny records) and they’re already deserialized my guess would be the serde of the public keys is still going to be the bigger perf hit. Since the hashmap’s gateways are already deserialized and you can pattern match the items you need I’d bet the `values` iterator is going to be roughly just as if not faster.

Reply:
But serde is not involved here is it ? we are going from a vec<u8> ( derived from a PublicKeyBinary.as_ref())  back to a PublicKeyBinary

Reply:
Not Serde the crate, but maybe you’re right about it going from a byte array to a pubkeybinary. I still worry about cloning 400k of those though

Reply:
See below for the code used for the revised implementation:

`        for gateway in self.gateway_cache_receiver.borrow().values() {
             if active_gateways.get(&gateway.address.as_ref().to_vec()).is_some() {
                if let Some(metadata) = &gateway.metadata {
                    global_map.increment_unclipped(metadata.location)
                }
            }
        }`